### ISSUE TITLE
Tuple support

### ISSUE THREAD

**Issue by colescott** [tarballs-are-good/coalton#15]
Coalton should support standard tuples

This can be achieved like with TYAPP and TYFUN using a new TYTUP type. The hope would be to allow for tuples in the language without having to explicitly specify every sized tuple constructor. 

Tuple decomposition in matches are also be a nice addition

**Comment by stylewarning** [comment]
I agree with this I think. Of course there won't be any functions that are generic on any sized tuple. (You'd have to specialize on tuple size.) But we can get support for the type expressions. We are 1/2 way doing this with `tyfun`s now which have arbitrary arity.
